File: cousins-in-binary-tree/cousins-in-binary-tree.cpp
/**
 * Definition for a binary tree node.
 * struct TreeNode {
 *     int val;
 *     TreeNode *left;
 *     TreeNode *right;
 *     TreeNode() : val(0), left(nullptr), right(nullptr) {}
 *     TreeNode(int x) : val(x), left(nullptr), right(nullptr) {}
 *     TreeNode(int x, TreeNode *left, TreeNode *right) : val(x), left(left), right(right) {}
 * };
 */
class Solution {
public:
    void parentAndDepth(TreeNode* root,int x,int y,int d,int& dx, int& dy, int& px, int& py){
        if(root==NULL) return;
        if(root->left){
            if(root->left->val==x){
                px=root->val; dx=d+1;
            }
            else if(root->left->val==y){
                py=root->val;
                dy=d+1;
            }
        }
        if(root->right){
            if(root->right->val==x){
                px=root->val; dx=d+1;
            }
            else if(root->right->val==y){
                py=root->val;
                dy=d+1;
            }
        }
        if(dx!=-1 and dy!=-1) return;
        parentAndDepth(root->left,x,y,d+1,dx,dy,px,py);
        parentAndDepth(root->right,x,y,d+1,dx,dy,px,py);
    }
    bool isCousins(TreeNode* root, int x, int y) {
        int px=0,py=0,dy=-1,dx=-1;
        parentAndDepth(root,x,y,0,dx,dy,px,py);
        return (dx==dy and px!=py);
    }
};

File: cousins-in-binary-tree/cousins-in-binary-tree.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public boolean isCousins(TreeNode root, int x, int y) {
        Queue<TreeNode> q=new LinkedList<TreeNode>();
        q.add(root);
        while(!q.isEmpty()){
            int n=q.size();
            boolean f1=false; boolean f2=false;
            for(int i=0;i<n;i++){
                TreeNode temp=q.remove();
                if(temp.val==x) f1=true;
                if(temp.val==y) f2=true;
                if(temp.left!=null && temp.right!=null){
                    if((temp.left.val==x && temp.right.val==y) || (temp.left.val==y && temp.right.val==x)) return false;
                }
                if(temp.left!=null) q.add(temp.left);
                if(temp.right!=null) q.add(temp.right);
            }
        if(f1==true && f2==true) return true;
        }
        return false;
    }
}